artificial_algorand_contract/trade/trade_contract.py

```diff
@@ -0,0 +1,224 @@
+from pyteal import (
+    And,
+    App,
+    Assert,
+    Bytes,
+    Concat,
+    Cond,
+    Ed25519Verify,
+    Global,
+    Gtxn,
+    If,
+    Int,
+    Mode,
+    OnComplete,
+    Return,
+    Seq,
+    compileTeal,
+)
+
+from ..classes.algorand import TealCmdList, TealPackage, TealParam
+from .asset_config import AssetConfig, aUSD_ID
+
+TEAL_VERSION = 5
+USD_ATOM_IN_ONE = Int(10**6)
+
+
+def gen_teal_param(asset_config: AssetConfig) -> TealParam:
+    """
+    Generate the teal param by AssetConfig
+    Args:
+        asset_config (AssetConfig): config of asset, can be found in asset_config.py
+    Returns:
+        TealParam: teal param
+    """
+    # for typing and readability
+
+    """ Smart contract typing """
+    cmd_list: TealCmdList = asset_config["contract_cmd_list"]
+    local_ints_scheme = asset_config["contract_local_ints_scheme"]
+    local_bytes_scheme = asset_config["contract_local_bytes_scheme"]
+    global_ints_scheme = asset_config["contract_global_ints_scheme"]
+    global_bytes_scheme = asset_config["contract_global_bytes_scheme"]
+
+    """ Dynamic Param from typing data """
+    teal_param: TealParam = {
+        "local_ints": len(local_ints_scheme),
+        "local_bytes": len(local_bytes_scheme),
+        "global_ints": len(global_ints_scheme),
+        "global_bytes": len(global_bytes_scheme),
+    }
+    return teal_param
+
+
+def approval_program(asset_config: AssetConfig) -> str:
+    """
+    Generate the approval contract TEAL code by AssetConfig.
+    Args:
+        asset_config (AssetConfig): config of asset, can be found in asset_config.py
+    Returns:
+        str: TEAL code of the approval contract
+    """
+    # All asset in this contract are counted in units of its own "decimal".
+    # for typing and readability
+    AAA_ID = asset_config["AAA_id"]  # raise exception?
+    if AAA_ID == 0:
+        input("AAA_ID shouldn't be 0")
+    PRICE_B16 = Int(int(asset_config["price"] * 2**16))
+    AAA_ATOM_IN_ONE = Int(10 ** asset_config["decimals"])
```

Comment:
clever setup! For price, it's going to be read from another contract (in next PRs)

Reply:
Yes. Algo this `PRICE_B16` should be `PRICE_B24`

Reply:
Correct! Also this `PRICE_B16` should be `PRICE_B24`. Added to DEV-107